### PR TITLE
#4591: Added ICU fields to PreviousHospitalization

### LINKED ISSUE #4591: Add Stay in ICU fields in previous hospitalizations [2]
### Bug Description
#4187 introduced additional Stay in ICU fields in previous hospitalizations. When entering previous hospitalizations, stay in ICU fields are not enabled.
### Steps to Reproduce
1. Go to cases. Go to hospitalizations.
2. Enter a previous hospitalization.

### Expected Behaviour
Stay in ICU fields should be enabled for previous hospitalizations.

**SPECIFICS:**
Cases > open case > Hospitalization tab > Previous Hospitalizations section > New Entry
Add "Stay in the intensive care unit?" as:
Yes/No/Unknown | If yes, show --> start date selector, end date selector

Related to https://github.com/hzi-braunschweig/SORMAS-Project/issues/1830 , as developed by @stefanspiska, but now applied to the new entries for "previous hospitalization" list.

image

### Screenshots
Screen Shot 2021-03-02 at 6 09 49 PM

### System Details
* Device:
* SORMAS version: 1.57.0
* Android version/Browser: chrome
* Server URL: test1401.sormas-oegd.de
* User Role: admin, national user

### Additional Information
https://github.com/hzi-braunschweig/sormas-survnet-interface/issues/31

### PR DESCRIPTION
**Pull request by ChristopherRiedel**
closes #4591

### CODE REVIEW

File: sormas-app/app/src/main/res/layout/dialog_previous_hospitalization_layout.xml [hunk 1]
```diff
@@ -61,6 +62,32 @@
...
+
+                <de.symeda.sormas.app.component.controls.ControlDateField
+                    android:id="@+id/caseHospitalization_intensiveCareUnitEnd"
+                    dependencyParentField="@{caseHospitalizationIntensiveCareUnit}"
+                    dependencyParentValue="@{YesNoUnknown.YES}"
+                    app:value="@={data.intensiveCareUnitEnd}"
+                    style="@style/ControlSecondOfTwoColumnsStyle" />
+
+            </LinearLayout>
+
```
Comment (DavidBaldsiefen) [inline sormas-app/app/src/main/res/layout/dialog_previous_hospitalization_layout.xml:90]: When I select ICU=Yes in the app, the datefields appearing below cannot be clicked (no date selector opens up). Can you replicate that?

**Review by DavidBaldsiefen — CHANGES_REQUESTED** [review]

File: sormas-app/app/src/main/res/layout/dialog_previous_hospitalization_layout.xml (hunk 1 above)
Comment (ChristopherRiedel) [reply sormas-app/app/src/main/res/layout/dialog_previous_hospitalization_layout.xml:90]: Yes I can reproduce this, but I don't have a clue what's the problem

Comment (DavidBaldsiefen) [reply sormas-app/app/src/main/res/layout/dialog_previous_hospitalization_layout.xml:90]: I also have no clue, sorry! Maybe something about the data binding didn't work?

**Review by DavidBaldsiefen — CHANGES_REQUESTED** [review]
Looks good so far, however in the android app the end date can be set before the start date.

I've fixed a similar problem in the PR for #4317

File: sormas-app/app/src/main/java/de/symeda/sormas/app/caze/edit/PreviousHospitalizationDialog.java [hunk 2]
```diff
@@ -12,10 +12,23 @@
...
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
```
Comment (DavidBaldsiefen) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/caze/edit/PreviousHospitalizationDialog.java:28]: This was probably unintended :)

File: sormas-app/app/src/main/java/de/symeda/sormas/app/caze/edit/CaseValidator.java [hunk 3]
```diff
@@ -214,7 +214,48 @@ static void initializePreviousHospitalizationValidation(final DialogPreviousHosp
...
+				}
+			}
+
+			return false;
+		};
+
 		contentBinding.casePreviousHospitalizationAdmissionDate.setValidationCallback(admissionDateCallback);
 		contentBinding.casePreviousHospitalizationDischargeDate.setValidationCallback(dischargeDateCallback);
+		contentBinding.casePreviousHospitalizationIntensiveCareUnitStart.setValidationCallback(intensiveCareUnitStartCallback);
+		contentBinding.casePreviousHospitalizationIntensiveCareUnitEnd.setValidationCallback(intensiveCareUnitEndCallback);
```
Comment (DavidBaldsiefen) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/caze/edit/CaseValidator.java:259]: Can you try using this instead as done in in `CaseEditFragment.java`?
```Java
ValidationHelper.initDateIntervalValidator(contentBinding.casePreviousHospitalizationIntensiveCareUnitStart, contentBinding.casePreviousHospitalizationIntensiveCareUnitEnd);
```

**Review by DavidBaldsiefen — CHANGES_REQUESTED** [review]

**Pull request merged**